### ISSUE TITLE
Plain SQL Templating Language undocumented use of colons (+ escaping)

### ISSUE THREAD

**Issue by juriad** [jOOQ/jOOQ#13477]
### Expected behavior

Plain SQL Templating Language is better documented and treatment of the colon is explicitly described, or treatment of the colon is removed. The language could also offer some escaping mechanism. The exact rules of parsing should be included.

### Actual behavior

Plain SQL treats the colon followed by name as a bind variable.

### Steps to reproduce the problem

I was trying to replicate this Postgres SQL:
```
postgres=# select a[:array_position(a, 3)-1] from (select '{1,2,3,4,5}'::int[] as a) b;
   a   
-------
 {1,2}
(1 row)
```

And found that
```
dsl.render(DSL.field("a[:array_position(a, {0})]", DSL.one()))
```
renders as:
```
a[1(a, 1)]
```

I can put space between `:` and `array_position` which solves it for me.

I could not find any mention of the colon being treated special anywhere. The only mentions involve question mark `?` and number or keyword in curly braces `{0}`, `{current_timestamp}`.
* https://www.jooq.org/doc/latest/manual/sql-building/plain-sql-templating/
* JavaDoc of org.jooq.SQL
* JavaDoc of org.jooq.impl.DSL#sql(java.lang.String, org.jooq.QueryPart...) - funnily this one refers to itself in `@see`

***

Another problematic situation would be support for some Postgres operators such as:
```
jsonb ? text → boolean
Does the text string exist as a top-level key or array element within the JSON value?
'{"a":1, "b":2}'::jsonb ? 'b' → t
'["a", "b", "c"]'::jsonb ? 'b' → t
```
https://www.postgresql.org/docs/14/functions-json.html
```
dsl.render(DSL.field("x ? {0}", DSL.one()))
```
renders as:
```
x 1 1
```

Is there any way to use this operator?

The operator `@?` is also rendered incorrectly `x @? {0}` as `x @1 1`.
The operator `?|` is rendered correctly  `x ?| {0}` as `x ?| 1`.

Since the user can define custom operators (https://www.postgresql.org/docs/14/sql-syntax-lexical.html#SQL-SYNTAX-OPERATORS and https://www.postgresql.org/docs/14/xoper.html), longer sequences of special characters can be an operator.
Trying `x ?? {0}` renders `x ?? 1`; `x ??? {0}` renders `x ??? 1`; `x ???? {0}` renders `x ???1 1`. Why does it work with 2 and 3 and breaks with 4?

***

I was about to suggest escaping `{?}` and `{:}` as doubling and backslash (`\`) do not look practical. And it even already works. This could be mentioned in the documentation/JavaDoc.
This also works as a consequence of the former: `dsl.render(DSL.field("x {0} {1}", DSL.keyword("?"), DSL.one()))`. The documentation of `Keyword` and `DSL.keyword(String)` does not mention that is can be used for escaping and operators, but the behavior looks consistent. 
Using a keyword also solves the original issue with `array_position`: `dsl.render(DSL.field("a[:{array_position}(a, {0})]", DSL.one()))`.

I love when I create an issue and find the answer myself. I still submit it because the documentation could be improved and others may find the workaround using keywords easier.

### Versions

- jOOQ: 3.16.5 pro
- Java: 11
- Database (include vendor): Postgres 11-14
- OS: Linux
- JDBC Driver (include name if inofficial driver):

**Comment by lukaseder** [comment]
Thanks a lot for your report. The `:` part is a documentation problem. Clearly, the plain SQL templating manual page should document this functionality (this is about named parameters).

Can you create a separate issue about the `?` part? It looks related, but it isn't. Escaping is a tricky thing to endeavour. Any new escaping syntax would likely lead to regressions in existing plain SQL templating code, so I'm reluctant to introduce it.

**Comment by juriad** [comment]
In the last section, I mention that escaping options already exist: `{?}` and `{:}`. It is not a new syntax but uses the existing documented way of rendering keywords. It is just a keyword without any letter in it. I would welcome a simple comment in the documentation that this is a valid solution. The keyword syntax even seems to work for curly braces: `a{{}b{{}c{}}d{}}e` -> `a{b{c}d}e`.

If we omit the escaping part, then what is left about `?` part is just documenting and possibly revising the rules of parsing. If a question mark is followed/preceded by one of these characters it is never considered a variable, for example.

There must be some list of special characters embedded in the grammar because `?+?` behaves differently from `?&?` and `?<?` behaves differently from `?>?`.

Playing a bit more with rendering, I can produce an ArrayIndexOutOfBoundsException probably due to some look-ahead? But that is a case that does not make much sense anyway.
![image](https://user-images.githubusercontent.com/886346/165310875-bef17fe6-366d-4091-8bb5-4e7d511f624a.png)

Maybe just mentioning that the user should always use whitespaces around the question mark or a small set of benign characters such as `()[]:,.` and explicitly use a keyword escaping for the operator that the question mark is part of, rather than [specifying the behavior](https://en.wikipedia.org/wiki/Unspecified_behavior) for various edge cases.

I can create an issue about the parsing rules of `?` and its inconsistencies. The rest is clear and can be documented as part of this issue?

**Comment by lukaseder** [comment]
> In the last section, I mention that escaping options already exist: `{?}` and `{:}`

Right. That isn't really an escaping syntax, though it can be used that way. You've already found out what it does, and it's actually documented in `org.jooq.SQL`'s Javadoc (look for keywords), just not in the manual.

> Maybe just mentioning that the user should always use whitespaces around the question mark or a small set of benign characters such as `()[]:,.` and explicitly use a keyword escaping for the operator that the question mark is part of, rather than [specifying the behavior](https://en.wikipedia.org/wiki/Unspecified_behavior) for various edge cases.

I'd rather the specification be predictable, or we break things incompatibly, inadvertedly in the future. The `:named` bind variable syntax has been supported for a long time and is compatible with Oracle or Spring, so it's just a documentation flaw. With this syntax in mind, it immediately follows that there can be conflicts with vendor specific syntax, just like the JDBC `?` marker can conflict with PostgreSQL syntax, too, and JDBC doesn't (yet) offer a standard way to escape it, though it has recently been discussed on the JDBC spec mailing list.

Anyway, I'll create new issues because this one is digressing a bit, and not every aspect might be addressed in the same release(s)

- The documentation flaw related to `:named` parameters: https://github.com/jOOQ/jOOQ/issues/13488
- The `AIOOBE`: https://github.com/jOOQ/jOOQ/issues/13489

> There must be some list of special characters embedded in the grammar because `?+?` behaves differently from `?&?` and `?<?` behaves differently from `?>?`.

Gee, why did they ever think those were good operator names 🙄

Anyway, some of the behaviour can be explained by e.g. #5307, #7035, #7037. But do know that even if you get past jOOQ's plain SQL templating, you will face similar limitations in the JDBC driver as well. I've documented some workarounds here: https://stackoverflow.com/q/38370972/521799

**Comment by lukaseder** [comment]
I've addressed #13488 and #13489. I'll be happy to look into more, separate, `?` related issues if you find any.

**Comment by lukaseder** [comment]
Closing this as a duplicate of the other two
